src/h3lib/lib/edge.c

```diff
@@ -0,0 +1,230 @@
+ */
+H3Error wrapDirectedEdgeError(H3Error err) {
+    if (err == E_DIR_EDGE_INVALID) {
+        return E_UNDIR_EDGE_INVALID;
+    }
+    return err;
+}
+
+/**
+ * Encode the edge between the two cells in non-normalized form.
+ * @param cell1 Origin cell
+ * @param cell2 Neighboring cell
+ * @param out Non-directed edge between the two cells, non-normalized
+ */
+H3Error cellsToEdgeNonNormalized(H3Index cell1, H3Index cell2, H3Index *out) {
+    // Determine the IJK direction from the origin to the destination
+    Direction direction = directionForNeighbor(cell1, cell2);
+
+    // The direction will be invalid if the cells are not neighbors
+    if (direction == INVALID_DIGIT) {
+        return E_NOT_NEIGHBORS;
+    }
+
+    // Create the edge index for the neighbor direction
+    H3Index output = cell1;
+    H3_SET_MODE(output, H3_EDGE_MODE);
+    H3_SET_RESERVED_BITS(output, direction);
+
+    *out = output;
+    return E_SUCCESS;
+}
+
+/**
+ * Allows for operations on an edge index as if it were a directed edge
+ * from the edge owner to the neighboring cell.
+ * @param edge Nondirected edge index
+ * @return H3Index Directed edge index
+ */
+H3Index edgeAsDirectedEdge(H3Index edge) {
+    // Do not make indexes that are not edges look "valid".
+    if (H3_GET_MODE(edge) == H3_EDGE_MODE) {
+        H3_SET_MODE(edge, H3_DIRECTEDEDGE_MODE);
+    }
+    return edge;
+}
+
+/**
+ * Normalize a nondirected edge.
+ *
+ * The normalization algorithm is that the owner of an edge
+ * is the cell with the numerically lower index.
+ * @param edge
+ * @param out
+ */
+H3Error normalizeEdge(H3Index edge, H3Index *out) {
+    H3Index originDestination[2] = {0};
+    H3Error odError = H3_EXPORT(edgeToCells)(edge, originDestination);
+    if (odError) {
+        return odError;
+    }
+    if (originDestination[1] < originDestination[0]) {
+        // The edge is not in normalized form already. Since there is only
+        // one other representation of this edge, we can be assured that
+        // reencoding with that representation will be normalized.
+        H3Error reencodeError = cellsToEdgeNonNormalized(
+            originDestination[1], originDestination[0], out);
+        if (reencodeError) {
+            return reencodeError;
+        }
+    } else {
+        // The edge is already in normalized form.
+        *out = edge;
+    }
+    return E_SUCCESS;
+}
+
+/**
+ * Returns an edge H3 index based on the provided neighboring cells
+ * @param cell1 An H3 hexagon index
+ * @param cell2 A neighboring H3 hexagon index
+ * @param out Output: the edge H3Index
+ */
+H3Error H3_EXPORT(cellsToEdge)(H3Index cell1, H3Index cell2, H3Index *out) {
+    H3Index nonNormalizedEdge;
+    H3Error nonNormalizedError =
+        cellsToEdgeNonNormalized(cell1, cell2, &nonNormalizedEdge);
+    if (nonNormalizedError) {
+        return nonNormalizedError;
+    }
+    H3Index normalizedEdge;
+    H3Error normalizeError = normalizeEdge(nonNormalizedEdge, &normalizedEdge);
```

Comment:
This seems potentially inefficient - I think the most expensive part here is `directionForNeighbor`, and we run it twice half the time. Why not just do
```
H3Index origin = cell1 < cell2 ? cell1 : cell2;
H3Index dest = cell1 < cell2 ? cell2 : cell1;
H3Error edgeErr = H3_EXPORT(cellsToDirectedEdge)(origin, dest, out);
H3_SET_MODE(out, H3_EDGE_MODE);
return edgeErr;
```
and drop the normalization post-processing?